Add support for sorting table with non-mutable columns

taldcroft opened the pull request:
Fixes #9536

EDIT: Close #9537

taldcroft commented on astropy/table/table.py:
```diff
@@ -2918,8 +2918,12 @@ def sort(self, keys=None, reverse=False):
+            try:
+                col[:] = new_col
+            except Exception:
+                self[col.info.name] = new_col
```
I tried setting to the new column in all cases but this caused problems with indexes which would require a deeper dive to understand / fix.

mhvk replied:
OK - I think here having a solution is definitely much better than worry about perfection. Maybe a new issue, though?

bsipocz commented on astropy/table/table.py:
```diff
@@ -2918,8 +2918,12 @@ def sort(self, keys=None, reverse=False):
+            new_col = col.take(indexes, axis=0)
+            try:
+                col[:] = new_col
+            except Exception:
```
could you make it to the specific exception? 
```suggestion
            except TypeError:
```

taldcroft replied:
This exception can be raised by arbitrary user-defined mixin columns, so I have no idea what exception class they might raise in this case.

bsipocz commented:
Looks good, one tiny comment on the exception

mhvk commented on astropy/table/tests/test_table.py:
```diff
@@ -2456,6 +2456,21 @@ class MyTable(table.Table):
+    assert np.allclose(t['sc'].ra.to_value(u.deg), [3, 4])
+    assert np.allclose(t['sc'].dec.to_value(u.deg), [5, 6])
+
+    # Got a deep copy of SkyCoord column
```
Are you sure you want to test this? This would break if eventually we change it so sorting can be done in-place. My sense is that all we want to test is that the `meta` are the same. Or were you thinking that the eventual goal should be anyway not to have this inplace, so this is desired behaviour that should stay this way? 

Thinking about this, I'm happy either way, but maybe make it explicit in the comment.

taldcroft replied:
I am happy enough if tests break later since that gives a clear indication of a behavior change, and then we understand what's going on and update the test.  These sort of tests were invaluable in all the monkeying around I did with meta copying.

mhvk replied:
OK, makes sense.

taldcroft commented:
I think I have addressed review comments.

@bsipocz - we need to do *something, anything* about the changelog conflict mess. :smiling_imp:  I just got hit with a conflict again.

taldcroft commented:
> Maybe a new issue, though?

#9553

bsipocz commented:
> we need to do something, anything about the changelog conflict mess.

This is really unfortunate, I agree. I'll try to play a bit with the rebase github action, it may just address the biggest painpoints we have.

(But just as with the json approach, one doesn't need to insert the new entry at the very bottom of the section 🤷‍♀ )

hamogu commented:
I thought new entries where supposed to be ordered by PR number. If that's not the case, things do get easier...

bsipocz commented:
No, there are no rules on ordering. Traditionally they are ordered by the order of merging, but that's that the same as the PR number, and the conflicts are coming from the fact that one usually adds the new entry at the bottom of a section and git is very careful to not auto-merge these, but flag them as conflict.  The rebase bot should easily handle it.
